<a href="https://colab.research.google.com/github/Vijaya381/sentiment-analysis-ann/blob/main/deep_learning1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip install opendatasets

In [ ]:
import os
os.environ["KAGGLE_API_TOKEN"]="kGAT_358de7a8b5a91ca54108ab8c6083e468"

In [ ]:
import opendatasets as od
od.download("https://www.kaggle.com/datasets/lakshmi25npathi/imdb-dataset-of-50k-movie-reviews")


Skipping, found downloaded files in "./imdb-dataset-of-50k-movie-reviews" (use force=True to force download)


In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns


In [ ]:
!ls


imdb-dataset-of-50k-movie-reviews  sample_data


In [ ]:
df=pd.read_csv("/content/imdb-dataset-of-50k-movie-reviews/IMDB Dataset.csv")

In [ ]:
df.head()

,review,sentiment
0,One of the other reviewers has mentioned that ...,positive
1,A wonderful little production. <br /><br />The...,positive
2,I thought this was a wonderful way to spend ti...,positive
3,Basically there's a family where a little boy ...,negative
4,"Petter Mattei's ""Love in the Time of Money"" is...",positive


In [ ]:
df.shape

(50000, 2)

In [ ]:
df.describe()

,review,sentiment
count,50000,50000
unique,49582,2
top,Loved today's show!!! It was a variety and not...,positive
freq,5,25000


In [ ]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review     50000 non-null  object
 1   sentiment  50000 non-null  object
dtypes: object(2)
memory usage: 781.4+ KB


50,000 reviews

 No missing values

 2 columns

 Binary classification problem

Text data (object)



In [ ]:
df["sentiment"].value_counts()

,count
sentiment,
positive,25000
negative,25000


In [ ]:
df["review"].str.len().describe()

,review
count,50000.000000
mean,1309.431020
std,989.728014
min,32.000000
25%,699.000000
50%,970.000000
75%,1590.250000
max,13704.000000


* count = 50000	There are 50,000 reviews
* mean = 1309	Average review length is about 1309 characters
* std = 989	Lengths vary a lot between reviews
* min = 32	Shortest review has 32 characters
* 25% = 699	25% of reviews are shorter than 699 characters
* 50% = 970	Median review length is 970 characters
* 75% = 1590	75% of reviews are shorter than 1590 characters
* max = 13704	Longest review has 13,704 characters

For your ANN project

The most important conclusions are:

* No missing values

* Balanced classes (25k positive, 25k negative)

* Plenty of training data

* Review lengths vary widely

Now lets move to preprocessing.

In [ ]:
df["sentiment"]=df["sentiment"].map({"positive":1,"negative":0})

In [ ]:
df.head()

,review,sentiment
0,One of the other reviewers has mentioned that ...,1
1,A wonderful little production. <br /><br />The...,1
2,I thought this was a wonderful way to spend ti...,1
3,Basically there's a family where a little boy ...,0
4,"Petter Mattei's ""Love in the Time of Money"" is...",1


**converting reviews into numbers**
bag of words

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
vectorizer = CountVectorizer(max_features=5000)
X = vectorizer.fit_transform(df["review"])


In [ ]:
print(X.shape)

(50000, 5000)


In [ ]:
print(len(vectorizer.vocabulary_))

5000


In [ ]:
print(list(vectorizer.vocabulary_.items())[:10])

[('one', np.int64(3121)), ('of', np.int64(3099)), ('the', np.int64(4442)), ('other', np.int64(3148)), ('reviewers', np.int64(3697)), ('has', np.int64(2067)), ('mentioned', np.int64(2822)), ('that', np.int64(4440)), ('after', np.int64(148)), ('watching', np.int64(4829))]


In [ ]:
feature_names = vectorizer.get_feature_names_out()

row = X[0].toarray()[0]

for idx, count in enumerate(row):
    if count > 0:
        print(feature_names[idx], count)

about 1
after 1
agenda 1
all 1
an 1
and 6
appeal 1
are 2
around 1
as 4
audiences 1
away 2
be 2
become 1
being 1
br 6
but 2
called 1
can 1
charm 1
city 2
class 1
classic 1
comfortable 1
couldn 1
dare 1
darker 1
death 1
developed 1
doesn 1
drugs 1
due 2
em 1
episode 2
ever 1
exactly 1
experience 1
experimental 1
face 1
fact 1
far 1
first 2
focuses 1
for 5
forget 3
from 1
get 2
given 1
glass 1
go 1
goes 1
got 1
graphic 1
guards 1
happened 1
hardcore 1
has 1
have 1
hearted 1
high 2
home 1
hooked 1
if 1
in 3
into 1
irish 1
is 9
it 6
its 2
just 2
kill 1
lack 1
levels 1
ll 3
main 1
mainly 1
mainstream 1
many 1
may 1
me 4
mentioned 1
mess 1
middle 1
more 2
nasty 1
never 1
no 1
not 3
of 7
on 3
one 1
or 3
order 1
other 2
out 1
oz 6
painted 1
pictures 1
pretty 1
prison 3
pulls 1
ready 1
reviewers 1
right 2
romance 1
saw 1
say 2
scenes 1
section 1
security 1
set 1
sex 1
show 3
shows 1
side 1
skills 1
so 3
sold 1
state 1
street 1
struck 2
surreal 1
taste 1
that 4
thats 1
the 16
their 1
they 1
thing

In [ ]:
y = df["sentiment"]

In [ ]:
print(X.shape)
print(y.shape)


(50000, 5000)
(50000,)


Train-Test Split

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

In [ ]:
print(X_train.shape)
print(X_test.shape)

print(y_train.shape)
print(y_test.shape)

(40000, 5000)
(10000, 5000)
(40000,)
(10000,)


In [ ]:
type(X_train)

scipy.sparse._csr.csr_matrix

convert to numpy array

In [ ]:
X_train = X_train.toarray()
X_test = X_test.toarray()

In [ ]:
print(type(X_train))
print(type(X_test))


<class 'numpy.ndarray'>
<class 'numpy.ndarray'>


In [ ]:
print(X_train.shape)
print(X_test.shape)

(40000, 5000)
(10000, 5000)


In [ ]:
import torch

X_train = torch.tensor(X_train, dtype=torch.float32)
X_test = torch.tensor(X_test, dtype=torch.float32)

y_train = torch.tensor(y_train.values, dtype=torch.float32)
y_test = torch.tensor(y_test.values, dtype=torch.float32)

In [ ]:
y_train = y_train.view(-1, 1)
y_test = y_test.view(-1, 1)

In [ ]:
print(y_train.shape)

torch.Size([40000, 1])


In [ ]:
print(X_train.shape)
print(y_train.shape)

torch.Size([40000, 5000])
torch.Size([40000, 1])


In [ ]:

import torch.nn as nn

class ANN(nn.Module):

    def __init__(self):
        super().__init__()

        self.fc1 = nn.Linear(5000, 128)
        self.fc2 = nn.Linear(128, 64)
        self.fc3 = nn.Linear(64, 1)

        self.relu = nn.ReLU()

    def forward(self, x):

        x = self.relu(self.fc1(x))
        x = self.relu(self.fc2(x))
        x = self.fc3(x)

        return x

In [ ]:
model = ANN()

print(model)

ANN(
  (fc1): Linear(in_features=5000, out_features=128, bias=True)
  (fc2): Linear(in_features=128, out_features=64, bias=True)
  (fc3): Linear(in_features=64, out_features=1, bias=True)
  (relu): ReLU()
)


In [ ]:
import torch.nn as nn

criterion = nn.BCEWithLogitsLoss()

outputs = model(X_train[:5])

print(outputs)
print(outputs.shape)

loss = criterion(outputs, y_train[:5])

print(loss)

tensor([[-0.0677],
        [-0.0481],
        [-0.0814],
        [-0.0728],
        [-0.0509]], grad_fn=<AddmmBackward0>)
torch.Size([5, 1])
tensor(0.6779, grad_fn=<BinaryCrossEntropyWithLogitsBackward0>)


In [ ]:
optimizer = torch.optim.Adam(
    model.parameters(),
    lr=0.001
)

optimizer.zero_grad()

loss.backward()

In [ ]:
print(model.fc1.weight.grad.shape)

torch.Size([128, 5000])


In [ ]:
print(model.fc1.weight.grad[:2, :5])

tensor([[ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000],
        [-0.0020,  0.0000,  0.0000,  0.0000,  0.0000]])


In [ ]:
epochs = 5

for epoch in range(epochs):

    optimizer.zero_grad()

    outputs = model(X_train)

    loss = criterion(outputs, y_train)

    loss.backward()

    optimizer.step()

    print(f"Epoch {epoch+1}, Loss: {loss.item():.4f}")

Epoch 1, Loss: 0.6929
Epoch 2, Loss: 0.6834
Epoch 3, Loss: 0.6698
Epoch 4, Loss: 0.6508
Epoch 5, Loss: 0.6292


In [ ]:
model.eval()

with torch.no_grad():

    outputs = model(X_test)

    probabilities = torch.sigmoid(outputs)

    predictions = (probabilities >= 0.5).float()

    accuracy = (predictions == y_test).float().mean()

print("Accuracy:", accuracy.item())

Accuracy: 0.7825999855995178
